full/runner.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2017 Palantir Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package full
+
+import (
+	"os"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/pkg/errors"
+
+	"github.com/palantir/bouncer/bouncer"
+)
+
+// Runner holds data for a particular full run
+// Note that in the full case, asgs will always be of length 1
+type Runner struct {
+	bouncer.BaseRunner
+}
+
+// NewRunner instantiates a new full runner
+func NewRunner(opts *bouncer.RunnerOpts) (*Runner, error) {
+	br, err := bouncer.NewBaseRunner(opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "error getting base runner")
+	}
+
+	r := Runner{
+		*br,
+	}
+	return &r, nil
+}
+
+// MustValidatePrereqs checks that the batch runner is safe to proceed
+func (r *Runner) MustValidatePrereqs() {
+	asgSet, err := r.NewASGSet()
+	if err != nil {
+		log.Fatal(errors.Wrap(err, "error building ASGSet"))
+	}
+
+	divergedASGs := asgSet.GetDivergedASGs()
+	if len(divergedASGs) != 0 {
+		for _, badASG := range divergedASGs {
+			log.WithFields(log.Fields{
+				"ASG":                     *badASG.ASG.AutoScalingGroupName,
+				"desired_capacity actual": *badASG.ASG.DesiredCapacity,
+				"desired_capacity given":  badASG.DesiredASG.DesiredCapacity,
+			}).Error("ASG desired capacity doesn't match expected starting value")
+		}
+		os.Exit(1)
+	}
+
+	for _, asg := range asgSet.ASGs {
+		if *asg.ASG.DesiredCapacity == 0 {
+			log.WithFields(log.Fields{
+				"ASG": *asg.ASG.AutoScalingGroupName,
+			}).Warn("ASG desired capacity is 0 - nothing to do here")
+			os.Exit(0)
+		}
+
+		if *asg.ASG.DesiredCapacity == *asg.ASG.MinSize {
+			log.WithFields(log.Fields{
+				"ASG":              *asg.ASG.AutoScalingGroupName,
+				"desired_capacity": *asg.ASG.DesiredCapacity,
+				"min_size":         *asg.ASG.MinSize,
+			}).Error("ASG desired capacity must be at least 1 higher than the min size, but they're equal")
```

Comment:
Actually, I _think_ what you want here is to enforce that min_size is 0. The idea is that we're taking all ASGs down to 0 before bringing them back, right?

Reply:
This is true.  The ASGs will need to be able to have their desired_capacity set to 0.

Reply:
ah ok, so we want that check to error out if ever *asg.ASG.MinSize > 0?

Reply:
updated.